File: Path Sum II.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
private:
    void pathSum(TreeNode* root, int target, vector<int>& path, vector<vector<int> >& result) {
        if (!root) 
            return;
        if (!root->left && !root->right && target == root->val) {
            path.push_back(root->val);
            result.push_back(path);
            path.pop_back();
            return;
        }
        path.push_back(root->val);
        pathSum(root->left, target - root->val, path, result);
        pathSum(root->right, target - root->val, path, result);
        path.pop_back();
    }
public:
    vector<vector<int> > pathSum(TreeNode *root, int sum) {
        vector<vector<int> > result;
        vector<int> path;
        pathSum(root, sum, path, result);
        return result;
    }
};

File: Path Sum II.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    private void findPath(TreeNode root, int target, List<Integer> path, List<List<Integer>> result) {
        if (root == null)
            return;
        if (root.left == null && root.right == null && root.val == target) {
            List<Integer> sol = new ArrayList<Integer>(path);
            sol.add(root.val);
            result.add(sol);
            return;
        }
        path.add(root.val);
        findPath(root.left, target - root.val, path, result);
        findPath(root.right, target - root.val, path, result);
        path.remove(path.size() - 1);
    }
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        List<Integer> path = new ArrayList<Integer>();
        findPath(root, sum, path, result);
        return result;
    }
}